File: maximum_product_subarray.py
""" 
Maximum Product Subarray
Given an integer array nums, find a subarray that has the largest product, and return the product.

Example :
Input: nums = [3,3,-2,4]
Output: 9
Explanation: [3,3] has the largest product 9.
"""

def maxProduct(nums):
    if not nums:
        return 0

    max_product = min_product = result = nums[0]

    for num in nums[1:]:
        if num < 0:
            max_product, min_product = min_product, max_product
        
        max_product = max(num, max_product * num)
        min_product = min(num, min_product * num)

        result = max(result, max_product)

    return result

print(maxProduct([3,3,-2,4]))

File: MaximumProductSubarray.java
public class MaximumProductSubarray {
    public int Solution(int nums[]){
        if(nums==null || nums.length==0){
            return 0;
        }
        int min=nums[0], max=nums[0];
        for(int i=1; i<nums.length; i++){
            int prevMin = min;
            int prevMax=max;
            min = Math.min(nums[i], Math.min(nums[i]*prevMin , nums[i]*prevMax));
            max = Math.max(nums[i] , Math.max(nums[i]*prevMin, nums[i]*prevMax));
        }
        return max;
    }

    public static void main(String[] args) {
        int arr1[]={5,3,1,-2,0,-5,8,-3};
        System.out.println(new MaximumProductSubarray().Solution(arr1));
    }
}
